File: carddemo/app/models/card.py
"""Card model — from CVACT02Y.cpy (150-byte VSAM KSDS records, key=CARD-NUM 16 chars).

COBOL record layout:
  05 CARD-NUM              PIC X(16)       → String(16) primary key
  05 CARD-ACCT-ID          PIC 9(11)       → BigInteger FK
  05 CARD-CVV-CD           PIC 9(03)       → Integer
  05 CARD-EMBOSSED-NAME    PIC X(50)       → String(50)
  05 CARD-EXPIRAION-DATE   PIC X(10)       → String(10)
  05 CARD-ACTIVE-STATUS    PIC X(01)       → String(1)
  05 FILLER                PIC X(59)       → not stored
"""

from sqlalchemy import BigInteger, ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class Card(Base):
    __tablename__ = "cards"

    card_num: Mapped[str] = mapped_column(String(16), primary_key=True)
    acct_id: Mapped[int] = mapped_column(BigInteger, ForeignKey("accounts.acct_id"))
    cvv_cd: Mapped[int] = mapped_column(Integer, default=0)
    embossed_name: Mapped[str] = mapped_column(String(50), default="")
    expiration_date: Mapped[str] = mapped_column(String(10), default="")
    active_status: Mapped[str] = mapped_column(String(1), default="Y")

File: demo/backend/src/main/java/com/carddemo/model/Card.java
package com.carddemo.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import jakarta.persistence.Version;

/**
 * JPA entity mapped from COBOL copybook CVACT02Y.cpy (Card Record, RECLN 150).
 *
 * COBOL layout:
 *   05 CARD-NUM               PIC X(16)
 *   05 CARD-ACCT-ID           PIC 9(11)
 *   05 CARD-CVV-CD            PIC 9(03)
 *   05 CARD-EMBOSSED-NAME     PIC X(50)
 *   05 CARD-EXPIRAION-DATE    PIC X(10)
 *   05 CARD-ACTIVE-STATUS     PIC X(01)
 */
@Entity
@Table(name = "cards")
public class Card {

    @Id
    @Column(name = "card_number", length = 16, nullable = false)
    private String cardNumber;

    @Column(name = "account_id", length = 11, nullable = false)
    private String accountId;

    // Equivalent to COBOL behavior: CVV is stored in VSAM record but never
    // displayed on any BMS screen (COCRDSL.bms / COCRDUP.bms).
    // @JsonIgnore prevents CVV from appearing in API responses (PCI DSS).
    @JsonIgnore
    @Column(name = "cvv_code", length = 3)
    private String cvvCode;

    @Column(name = "embossed_name", length = 50)
    private String embossedName;

    @Column(name = "expiration_date", length = 10)
    private String expirationDate;

    @Column(name = "active_status", length = 1)
    private String activeStatus;

    // Equivalent to 9300-CHECK-CHANGE-IN-REC in COCRDUPC.cbl
    // COBOL checks all field values before/after to detect concurrent modification.
    // JPA @Version provides optimistic locking — throws OptimisticLockException
    // if another transaction modified the record between read and write.
    @Version
    @Column(name = "version")
    private Long version;

    public Card() {}

    public Card(String cardNumber, String accountId, String cvvCode,
                String embossedName, String expirationDate, String activeStatus) {
        this.cardNumber = cardNumber;
        this.accountId = accountId;
        this.cvvCode = cvvCode;
        this.embossedName = embossedName;
        this.expirationDate = expirationDate;
        this.activeStatus = activeStatus;
    }

    public String getCardNumber() { return cardNumber; }
    public void setCardNumber(String cardNumber) { this.cardNumber = cardNumber; }

    public String getAccountId() { return accountId; }
    public void setAccountId(String accountId) { this.accountId = accountId; }

    public String getCvvCode() { return cvvCode; }
    public void setCvvCode(String cvvCode) { this.cvvCode = cvvCode; }

    public String getEmbossedName() { return embossedName; }
    public void setEmbossedName(String embossedName) { this.embossedName = embossedName; }

    public String getExpirationDate() { return expirationDate; }
    public void setExpirationDate(String expirationDate) { this.expirationDate = expirationDate; }

    public String getActiveStatus() { return activeStatus; }
    public void setActiveStatus(String activeStatus) { this.activeStatus = activeStatus; }

    public Long getVersion() { return version; }
    public void setVersion(Long version) { this.version = version; }
}
